site/bps/BP-35-128-bits-support.md

```diff
@@ -0,0 +1,399 @@
+
+##### Entry Format V2
+
+```json
+Entry Format V2
+===============
+--- header ---
+Bytes (0 - 7)                   : Metadata Flags
+Bytes (8 - 15)                  : Ledger Scope ID
+Bytes (16 - 23)                 : Ledger ID
+Bytes (24 - 31)                 : Entry ID
+Bytes (32 - 39)                 : LastAddConfirmed
+Bytes (40 - 47)                 : Length
+--- digest ---
+Bytes (37 - (37 + x - 1))       : Digest Code (e.g. CRC32)
+--- payload ---
+Bytes ((37 + x) - )             : Payload
+``` 
+
+> `x` is the length of digest code.
+
+###### Metadata Flags
+
+```json
+Metadata: 1 Bytes (Long)
+------------------------
+0x 0 0
+   |__| 
+     |
+ version
+
+----
+Bit 0 - 3: digest type (e.g. CRC32, CRC32C and such)
+Bit 4 - 7: version, the most significant bit of this byte will be always set to 1.
+it will be used for differentiating entry format v1 and v2.
+
+```
+
+We are setting the most significant bit to be `1`. So the first byte in entry v2 will
+be a negative value, which can be used for differentiating entry format v1 and v2.
+The version will be encoded into the first byte. The version will be used for describing
+the entry format.
+
+##### Decoding Entry
+
+The pseudo code for decoding an entry will be described as followings:
+
+```java
+
+ByteBuf entry = ...;
+
+int metadataFlags = entry.getByte();
+
+if (metadataFlags <= 128) { // the entry is encoded in v1 format
+	// decoding the entry in v1 format
+	...
+} else {
+	// decoding the entry in v2 format
+}
+
+```
+
+#### Bookie Storage
+
+##### Journal
+
+A new method should be added in journal `WriteCallback` to handle `ledgerScopeId`.
+
+```java
+public interface WriteCallback {
+
+    void writeComplete(int rc,
+                       long ledgerScopeId,
+                       long ledgerId,
+                       long entryId,
+                       BookieSocketAddress addr,
+                       Object ctx);
+
+    default void writeComplete(int rc,
+                               long ledgerId,
+                               long entryId,
+                               BookieSocketAddress addr,
+                               Object ctx) {
+        writeComplete(rc, 0L, ledgerId, entryId, addr, ctx);
+    }
+
+}
+```
+
+The journal should be changed to be able to retrieve `ledgerScopeId` from the entry
+payload based on [Entry Format](#entry-format).
+
+##### Ledger Storage
+
+###### EntryLogger
+
+1. Methods in `EntryLogger` should be able to accept `ledgerScopeId` as a parameter.
+2. EntryLogger should be updated to retrieve `ledgerScopeId` from the entry payload
+   based on [Entry Format](#entry-format).
+
+###### EntryMemTable
+
+`ledgerScopeId` should be added as part of `EntryKey`.
+
+###### IndexPersistenceMgr
+
+Currently the ledger index files (64 bits) are stored into 2-level-hirechicy
+directories - `<msb-32bits-hex>/<lsb-32bits-hex>/<ledger-id-hex>.idx`.
+
+If `ledgerScopeId` is 0, it will be using existing scheme for storing and retrieving
```

Comment:
Why bifurcate the code paths like this? ledgerScopeId 0 should be treated as special in as few places as possible.

Reply:
This is for BC consideration. You already have a ledger manager running with ledgerScopeId == 0. you have to keep the logic functioning as it is needed.

Reply:
I'd prefer a migration script for this rather than keeping two independent code paths running at the same time. We need to keep the old path for people not using scopeId anyhow. the migration itself is easy (just a mv). If a bookie is going to be aware of the scopeId they need to upgrade their journal anyhow, so this could be part of the same upgrade.

Reply:
> the migration itself is easy (just a mv).

a migration script can be problematic for a production system. it can become much complicated than keeping BC. 

I think we can start a brand new metadata driver as I commented in mailing list for Enrico's question, for existing metadata driver keeping the BC and maybe providing a migration script for people to cutover if needed.

Reply:
I am not sure I understand your point here. 

Upgrading Journal is much easier as it doesn't require any coordination. However upgrading metadata is not, it includes coordination between clients (reader and writer) and bookies. A script on migrating metadata doesn't really work out in a production system.

Reply:
oh sorry I misunderstood your comment. then a migration/upgrade script works